crates/ironrdp-client-blocking/src/rdp.rs

```diff
@@ -0,0 +1,270 @@
+use crate::config::Config;
+use ironrdp::connector::{ConnectionResult, ConnectorResult};
+use ironrdp::graphics::image_processing::PixelFormat;
+use ironrdp::pdu::input::fast_path::FastPathInputEvent;
+use ironrdp::session::image::DecodedImage;
+use ironrdp::session::{ActiveStage, ActiveStageOutput, SessionResult};
+use ironrdp::{connector, session};
+use smallvec::SmallVec;
+use sspi::network_client::reqwest_network_client::RequestClientFactory;
+use std::io;
+use std::net::TcpStream;
+use std::sync::Arc;
+use winit::event_loop::EventLoopProxy;
+
+#[derive(Debug)]
+pub enum RdpOutputEvent {
+    Image { buffer: Vec<u32>, width: u16, height: u16 },
+    ConnectionFailure(connector::ConnectorError),
+    Terminated(SessionResult<()>),
+}
+
+#[derive(Debug)]
+pub enum RdpInputEvent {
+    Resize {
+        width: u16,
+        height: u16,
+    },
+    FastPath(SmallVec<[FastPathInputEvent; 2]>),
+    Close,
+    /// DevNull is a special event that is used to check if the input event channel is closed.
+    DevNull,
+}
+
+impl RdpInputEvent {
+    pub fn create_channel() -> (crossbeam::channel::Sender<Self>, crossbeam::channel::Receiver<Self>) {
+        crossbeam::channel::unbounded()
+    }
+}
+
+pub struct RdpClient {
+    pub config: Config,
+    pub event_loop_proxy: EventLoopProxy<RdpOutputEvent>,
+    pub input_event_receiver: crossbeam::channel::Receiver<RdpInputEvent>,
+}
+
+impl RdpClient {
+    pub fn run(mut self) {
+        loop {
+            let (connection_result, framed) = match connect(&self.config) {
+                Ok(result) => result,
+                Err(e) => {
+                    let _ = self.event_loop_proxy.send_event(RdpOutputEvent::ConnectionFailure(e));
+                    break;
+                }
+            };
+
+            match active_session(
+                framed,
+                connection_result,
+                &self.event_loop_proxy,
+                &mut self.input_event_receiver,
+            ) {
+                Ok(RdpControlFlow::ReconnectWithNewSize { width, height }) => {
+                    self.config.connector.desktop_size.width = width;
+                    self.config.connector.desktop_size.height = height;
+                }
+                Ok(RdpControlFlow::TerminatedGracefully) => {
+                    let _ = self.event_loop_proxy.send_event(RdpOutputEvent::Terminated(Ok(())));
+                    break;
+                }
+                Err(e) => {
+                    let _ = self.event_loop_proxy.send_event(RdpOutputEvent::Terminated(Err(e)));
+                    break;
+                }
+            }
+        }
+    }
+}
+
+enum RdpControlFlow {
+    ReconnectWithNewSize { width: u16, height: u16 },
+    TerminatedGracefully,
+}
+
+type UpgradedFramed = ironrdp_blocking::Framed<native_tls::TlsStream<TcpStream>>;
```

Comment:
I went with `native_tls::TlsStream` for now, surely it can also be done with `rustls`.

Reply:
Definitely! Will look into that later